rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,122 @@
+- Providing a new "safe" Domain Specific Language for IaSQL: developers can use this language to create new mappers, and we can translate that to TypeScript code on-air, while narrowing down the DSL interface to something that is "safe" (no code-execution, no unwanted database access, etc). The DSL does not have to be capable of creating all possible mappers (at least not at first) since if they want to do something very professional, they can just submit a PR to the engine.
+  - For example: we can supply a base provider like AWS. And using DSL people can dynamically invoke AWS client's methods and write their logic. The DSL then gets parsed and converted to JavaScript code, and we'll use its interface when we need the mapper.
+
+
+#### Mapper Logic Inside Postgres
+
+We can put the mapper's logic inside Postgres. And then provide an interface for the engine to call mapper functions. In this approach, the interface will be provided through the Postgres, and therefore this is the approach that is most consistent with "Pure-SQL". The DSL idea from the above way can still be applied here (developer writes DSL, PL/pgSQL code is generated).
+
+So let's consider one possible implementation of the mapper structure (all as Postgres functions):
+- CloudId: `{entity}_cloud_id(object: json) -> string` a Postgres function that takes the object as json and returns its cloud id as string.
+- Equals: `{entity}_equal(o1: json, o2: json) -> boolean`
+- CRUD functions:
+    - `{entity}_cloud_create(object: json)`
+    - `{entity}_cloud_read() -> json list`
+    - `{entity}_cloud_update(old: json, new: json)`
+    - `{entity}_cloud_delete(object: json)`
+- Call other mappers: 
+  - `call_mapper('{module_name}', '{entity_name}', 'cloud|db', 'create|read|update|delete', {}::json) -> json`
+  - It would be easy to add support for normal modules with the above definition. It will call `ec2_cloud_read` if that Postgres function exists. Otherwise, it'll send an RPC to the engine to ask engine to call that function from the mapper.
+
+We also need some other things for this to work:
+- Dynamically creating a module that provides the same interface as `MapperBase`, but calls corresponding SQL functions. 
+
+### Alternatives Considered
+
+I also considered the functions to be strongly-typed inside Postgres. More info here:
+https://rounded-apology-58d.notion.site/What-Does-a-Mapper-Need-9402ce7e35244405876f4cf660365c86
+
+## Expected Semver Impact
+
+A minor update. Just adding some Postgres functions, the ability to install a Pure-SQL module from remote link, and the engine having the ability to work with Pure-SQL modules. 
+
+## Affected Components
+
+Mostly the engine
+
+## Expected Timeline
+
+I believe it takes at least 2 weeks for us to create a working example of this and port an easy module to it.
```

Comment:
I think we could get something quicker than 2 weeks if we're okay with cut corners for the demonstration that we then clean up.